core/src/main/java/hudson/model/Run.java

```diff
@@ -2426,6 +2426,12 @@ public EnvVars getEnvironment() throws IOException, InterruptedException {
      */
     public @NonNull EnvVars getEnvironment(@NonNull TaskListener listener) throws IOException, InterruptedException {
         Computer c = Computer.currentComputer();
+
+        if (c == null) {
+            Executor e = this.getExecutor();
+            c = e == null ? null : e.owner;
```

Comment:
For Pipeline, this will be the controller. And we do not want to pass controller environment variables to the build necessarily.

Reply:
You are right - I misunderstood the handling of Pipelines in this context. I didn't notice it because I had not attached agents to my local dev instance. Thanks!